ethcore/src/client/client.rs

```diff
@@ -1511,6 +1511,49 @@ impl CallContract for Client {
 			.map_err(|e| format!("{:?}", e))
 			.map(|executed| executed.output)
 	}
+
+	fn call_contract_before(&self, header: &Header, address: Address, data: Bytes) -> Result<Bytes, String> {
```

Comment:
Why is that even needed? Why can't use use `call_contract` and pass `BlockId` with `parent_hash`?

Reply:
We can definitely use `call` later on in this function. I'm not sure about replacing `call_contract_before` with `call_contract` because `state` is defined differently in those two functions.

Reply:
It looks like the use of `mut state` in `call_contract_before` is incorrect because it is then dropped. So whatever changes `call` makes to it are forgotten.

@afck, what are your thoughts on using `call_contract` instead of `call_contract_before` (which should fix the use of mutable state)?

Reply:
You're probably right: We should just pass in the block number into `block_gas_limit` as a parameter.

Originally this was motivated by a smart contract (written before this Rust code) that had a [`limitBlockGas` getter](https://github.com/poanetwork/posdao-contracts/blob/3b5db5f74082da96c161dec0010105078d2f4b81/contracts/TxPermission.sol#L181) that returned whether the gas limit should be reduced for the _current_ block. So we needed a (["not exactly sane"](https://github.com/poanetwork/parity-ethereum/pull/140#issue-284967844)) way to call this with a state that didn't exist yet: where the block number is the number of the block that we're currently creating. The [new `blockGasLimit` implementation](https://github.com/poanetwork/posdao-contracts/blob/2f3449b765494af346503a3de52f0fdfbe425eef/contracts/TxPermission.sol#L223) has a similar problem: It calls a few methods that use the current block number.
@varasev: Would it be feasible to just pass the block number into `blockGasLimit` as an argument instead?

> It looks like the use of `mut state` in `call_contract_before` is incorrect because it is then dropped.

[`Client::call_contract`](https://github.com/paritytech/parity-ethereum/blob/3420c2bac08353f4269b9589decf8a9351275e64/ethcore/src/client/client.rs#L1401) drops the state as well: It ends up [via `call`](https://github.com/paritytech/parity-ethereum/blob/3420c2bac08353f4269b9589decf8a9351275e64/ethcore/src/client/client.rs#L1482) in [`do_virtual_call`](https://github.com/paritytech/parity-ethereum/blob/3420c2bac08353f4269b9589decf8a9351275e64/ethcore/src/client/client.rs#L1169), where it is used to [create an `Executive` instance](https://github.com/paritytech/parity-ethereum/blob/3420c2bac08353f4269b9589decf8a9351275e64/ethcore/src/client/client.rs#L1194) and a diff. It's not persisted, however, and that's intended: These methods are meant to run calls _locally_ and return the call's output. They don't create transactions and thus can't modify the actual EVM state.

Reply:
> Would it be feasible to just pass the block number into `blockGasLimit` as an argument instead?

No, the `blockGasLimit` calls other getters of other contracts to determine the gas limit for the current block. The returned values of those getters depend on the current state of the current block.